### PR TITLE
Add Write Models Docs

### PR DESCRIPTION
**Pull request by rentainhe**
## TODO
- [x] 添加`Write Models`相关的Tutorials
- [x] 修复一些文档的错误

这部分可能需要分成几个part写

### CODE REVIEW

**Comment by rentainhe** [comment]
## 修复layers的文档错误
- 修复前:
image

- 修复后:
image

**Comment by rentainhe** [comment]
## 新增的layers接口文档
image

**Comment by rentainhe** [comment]
## 添加 Write Models 的教程
image

File: docs/source/tutorials/Write_Models.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Write Models
+
+Here we provides the tutorials for users who want to implement new model entirely from scratch and make it compatible in LiBai.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Write_Models.md:3]:
```suggestion
In this section, we are showing users of how to implement new model entirely from scratch and make it compatible with LiBai.
```

Comment (Ldpe2G) [reply docs/source/tutorials/Write_Models.md:3]: 这样如何

File: docs/source/tutorials/Write_Models.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+
+## Construct Models in LiBai
+
+LiBai uses ``LazyConfig`` for more flexible config system, which means you can simply import your own model in your config and train it under LiBai.
+
+For example, in image classification task, the input data is usually a batch of images with its targets during training and only input a batch of images when testing, to build a toy model for image classification task, import this code in your own code:
```
Comment (Ldpe2G) [inline docs/source/tutorials/Write_Models.md:10]:
```suggestion
For image classification task, the input data is usually a batch of images and labels during training, and only images when testing. The following code shows how to build a toy model for image classification task:
```

File: docs/source/tutorials/Write_Models.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        if labels is not None and self.training:
+            losses = self.loss_func(x, labels)
+            return {"losses": losses}
+        else:
+            return {"prediction_scores": x}
+```
+
+**Note:**
+- For classification models, the ``forward`` function of each model must have ``images`` and ``labels`` as arguments, which corresponds to the output of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Write_Models.md:42]:
```suggestion
- For classification models, the ``forward`` function must have ``images`` and ``labels`` as arguments, which corresponds to the output of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.
```

File: docs/source/tutorials/Write_Models.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+- Each model returns ``losses`` during training and returns ``prediction_scores`` during inference, and both of them should be the type of ``dict``, which means you should implement the ``loss function`` in your model, like ``self.loss_func=nn.CrossEntropyLoss()`` as the ToyModel.
+
+
+## Import the model in config
+
+With ``LazyConfig System``, you can simply import the model in your config file to use ``ToyModel`` for training as follows:
```
Comment (Ldpe2G) [inline docs/source/tutorials/Write_Models.md:48]:
```suggestion
With ``LazyConfig System``, you can simply import the model in your config file. The following code shows how to use ``ToyModel`` for training:
```

File: docs/source/tutorials/Write_Models.md (hunk 3 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md]: 已修改~

File: docs/source/tutorials/Write_Models.md (hunk 4 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md]: 已修改~

File: docs/source/tutorials/Write_Models.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        if labels is not None and self.training:
+            losses = self.loss_func(x, labels)
+            return {"losses": losses}
+        else:
+            return {"prediction_scores": x}
+```
+
+**Note:**
+- For classification models, the ``forward`` function of each model must have ``images`` and ``labels`` as arguments, which corresponds to the output of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.
+- Each model returns ``losses`` during training and returns ``prediction_scores`` during inference, and both of them should be the type of ``dict``, which means you should implement the ``loss function`` in your model, like ``self.loss_func=nn.CrossEntropyLoss()`` as the ToyModel.
```
[the remark this answers is not in the document]
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md]: 已修改~

File: docs/source/tutorials/Write_Models.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
+# Write Models
+
+In this section, we are showing users how to implement new model entirely from scratch and make it compatible with LiBai.
```
Comment (CPFLAME) [inline docs/source/tutorials/Write_Models.md:3]:
```suggestion
In this section, we will introduce how to implement new model entirely from scratch and make it compatible with LiBai.
```

File: docs/source/tutorials/Write_Models.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+
+## Construct Models in LiBai
+
+LiBai uses ``LazyConfig`` for more flexible config system, which means you can simply import your own model in your config and train it under LiBai.
```
Comment (CPFLAME) [inline docs/source/tutorials/Write_Models.md:8]: 这里的lazyconfig可以给个 超链接到lazyconfig的doc说明

File: docs/source/tutorials/Write_Models.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+For image classification task, the input data is usually a batch of images and labels during training, and only images when testing. The following code shows how to build a toy model for image classification task, import this code in your code:
```
Comment (CPFLAME) [inline docs/source/tutorials/Write_Models.md:10]: 这里可能不太准确, 因为在libai里面  在`testing`的时候 也是有`labels`的

File: libai/tokenizer/tokenization_base.py [hunk 9]
```diff
@@ -183,6 +191,7 @@ def from_pretrained(cls, *inputs, **kwargs):
         r"""
         Instantiate a :class:`~transformers.PreTrainedTokenizer` (or a derived class) from a
         predefined tokenizer.
+        
         Args:
             pretrained_model_name_or_path: either:
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:196]: https://github.com/huggingface/transformers/blob/aa6cfe9c4b073b2c058a78fc2d26fe3fbe0ad70b/src/transformers/tokenization_utils_base.py#L1553 can be either

File: libai/tokenizer/tokenization_base.py [hunk 10]
```diff
@@ -406,16 +419,19 @@ def _from_pretrained(cls, pretrained_model_name_or_path, *init_inputs, **kwargs)
...
-        """Save the tokenizer vocabulary files together with:
+        """
+        Save the tokenizer vocabulary files together with:
+            
             - added tokens,
             - special-tokens-to-class-attributes-mapping,
             - tokenizer instantiation positional and keywords inputs
             (e.g. do_lower_case for Bert).
+        
         This won't save modifications other than (added tokens and special token mapping)
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:430]: This -> This method

Comment (thinksoso) [reply libai/tokenizer/tokenization_base.py:430]: (added tokens and special token mapping) 应该不用括号括起来

Comment (rentainhe) [reply libai/tokenizer/tokenization_base.py:430]: OK, 这个部分的表达其实我没改过，估计是从哪儿copy过来的2333，所以可能需要校对

File: docs/source/tutorials/Write_Models.md (hunk 8 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md:10]: OK，我来改一下

File: docs/source/tutorials/Write_Models.md (hunk 7 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md:8]: 好，OK

File: libai/tokenizer/tokenization_base.py [hunk 11]
```diff
@@ -497,22 +515,29 @@ def add_tokens(self, new_tokens: Union[str, List[str]], special_tokens: bool = F
         Add a list of new tokens to the tokenizer class. If the new tokens are not in the
         vocabulary, they are added to it with indices starting from length of
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:516]: from length -> the length

File: libai/tokenizer/tokenization_base.py [hunk 12]
```diff
@@ -497,22 +515,29 @@ def add_tokens(self, new_tokens: Union[str, List[str]], special_tokens: bool = F
...
         the current vocabulary.
+        
         .. Note::
             When adding new tokens to the vocabulary, you should make sure to also resize
             the token embedding matrix of the model so that its embedding matrix matches
             the tokenizer.
             In order to do that, please use the
             :meth:`~transformers.PreTrainedModel.resize_token_embeddings` method.
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:524]: ~~transformers.~~

File: libai/tokenizer/tokenization_base.py [hunk 13]
```diff
@@ -497,22 +515,29 @@ def add_tokens(self, new_tokens: Union[str, List[str]], special_tokens: bool = F
         Add a list of new tokens to the tokenizer class. If the new tokens are not in the
         vocabulary, they are added to it with indices starting from length of
         the current vocabulary.
+        
         .. Note::
             When adding new tokens to the vocabulary, you should make sure to also resize
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:520]: you should make sure to also resize -> you should also make sure to resize

File: libai/tokenizer/tokenization_base.py [hunk 14]
```diff
@@ -577,11 +602,13 @@ def add_special_tokens(self, special_tokens_dict: Dict[str, str]) -> int:
         Add a dictionary of special tokens (eos, pad, cls, etc.) to the encoder and link them to
         class attributes. If special tokens are NOT in the vocabulary, they are added to it
         (indexed starting from the last index of the current vocabulary).
+        
         .. Note::
             When adding new tokens to the vocabulary, you should make sure to also resize the
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:607]: you should make sure to also resize -> you should also make sure to resize

File: libai/tokenizer/tokenization_base.py [hunk 15]
```diff
@@ -577,11 +602,13 @@ def add_special_tokens(self, special_tokens_dict: Dict[str, str]) -> int:
...
             token embedding matrix of the model so that its embedding matrix matches the tokenizer.
             In order to do that, please use the
             :meth:`~transformers.PreTrainedModel.resize_token_embeddings` method.
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:610]: ~~transformers~~

File: libai/tokenizer/tokenization_base.py [hunk 16]
```diff
@@ -590,6 +617,7 @@ class attributes. If special tokens are NOT in the vocabulary, they are added to
         When possible, special tokens are already registered for provided pretrained models
         (for instance :class:`~transformers.BertTokenizer` :obj:`cls_token` is already registered
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:618]: ~~transformers~~

File: libai/tokenizer/tokenization_base.py [hunk 17]
```diff
@@ -639,12 +672,14 @@ def tokenize(self, text: str, **kwargs) -> List[str]:
         Converts a string in a sequence of tokens, using the tokenizer.
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:672]: in -> to

File: libai/tokenizer/tokenization_base.py [hunk 18]
```diff
@@ -639,12 +672,14 @@ def tokenize(self, text: str, **kwargs) -> List[str]:
...
         Split in words for word-based vocabulary or sub-words for sub-word-based vocabularies
         (BPE/SentencePieces/WordPieces). Takes care of added tokens.
```
Comment (thinksoso) [inline libai/tokenizer/tokenization_base.py:674]: Takes -> Take

File: libai/tokenizer/tokenization_base.py (hunk 13 above)
Comment (thinksoso) [reply libai/tokenizer/tokenization_base.py:520]: 不用改这个 : - )

File: libai/tokenizer/tokenization_base.py (hunk 10 above)
Comment (thinksoso) [reply libai/tokenizer/tokenization_base.py:430]: 应该是改的transformers，不过他们的原生文档有些地方感觉也不是很通顺，但是考虑他们是native speaker还是不动了，只把一些后来新加的错误纠正一下

Comment (rentainhe) [reply libai/tokenizer/tokenization_base.py:430]: 原来如此

File: docs/source/tutorials/Write_Models.md (hunk 6 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md]: 已改~

File: docs/source/tutorials/Write_Models.md (hunk 8 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md]: 我改了~再看看呢~

File: libai/tokenizer/tokenization_base.py (hunk 9 above)
Comment (rentainhe) [reply libai/tokenizer/tokenization_base.py]: OK，这里已经修改了

File: libai/tokenizer/tokenization_base.py (hunk 12 above)
Comment (rentainhe) [reply libai/tokenizer/tokenization_base.py]: 已改

File: libai/tokenizer/tokenization_base.py (hunk 16 above)
Comment (rentainhe) [reply libai/tokenizer/tokenization_base.py]: 已改

File: libai/tokenizer/tokenization_base.py (hunk 15 above)
Comment (rentainhe) [reply libai/tokenizer/tokenization_base.py]: 已改~

File: libai/tokenizer/tokenization_base.py (hunk 11 above)
Comment (rentainhe) [reply libai/tokenizer/tokenization_base.py]: 已改~

File: libai/tokenizer/tokenization_base.py (hunk 10 above)
Comment (rentainhe) [reply libai/tokenizer/tokenization_base.py]: 已修复

File: docs/source/tutorials/Write_Models.md [hunk 19]
```diff
@@ -0,0 +1,60 @@
+# Write Models
+
+In this section, we will introduce how to implement new model entirely from scratch and make it compatible with LiBai.
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Write_Models.md:3]:
```suggestion
In this section, we will introduce how to implement a new model entirely from scratch and make it compatible with LiBai.
```

File: docs/source/tutorials/Write_Models.md [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+
+
+## Construct Models in LiBai
+
+LiBai uses [LazyConfig](https://libai.readthedocs.io/en/latest/tutorials/Config_System.html) for more flexible config system, which means you can simply import your own model in your config and train it under LiBai.
+
+For image classification task, the input data is usually a batch of images and labels. The following code shows how to build a toy model for image classification task, import this code in your code:
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Write_Models.md:10]:
```suggestion
For image classification task, the input data is usually a batch of images and labels. The following code shows how to build a toy model for this task, import it in your code:
```

File: docs/source/tutorials/Write_Models.md [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+```python
+# toy_model.py
+import oneflow as flow
+import oneflow.nn as nn
+
+
+class ToyModel(nn.Module):
+    def __init__(self, 
+        num_classes=1000, 
+        loss_func=None
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Write_Models.md:20]: 这里 loss_func 作为入参，并没有用到

File: docs/source/tutorials/Write_Models.md [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+        if labels is not None and self.training:
+            losses = self.loss_func(x, labels)
+            return {"losses": losses}
+        else:
+            return {"prediction_scores": x}
+```
+
+**Note:**
+- For classification models, the ``forward`` function must have ``images`` and ``labels`` as arguments, which corresponds to the output of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Write_Models.md:42]:
```suggestion
- For classification models, the ``forward`` function must have ``images`` and ``labels`` as arguments, which corresponds to the output in ``getitem`` of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.
```

File: docs/source/tutorials/Write_Models.md [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+- Each model will return ``losses`` during training and ``prediction_scores`` during inference, both of which should be the type of ``dict``, which means you should implement the ``loss function`` in your model, like ``self.loss_func=nn.CrossEntropyLoss()`` as the ToyModel showing above.
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Write_Models.md:43]:
```suggestion
- This toy model will return ``losses`` during training and ``prediction_scores`` during inference, and both of them should be the type of ``dict``, which means you should implement the ``loss function`` in your model, like ``self.loss_func=nn.CrossEntropyLoss()`` as the ToyModel showing above.
```

这里强调一下是这个 example 里面的 toy model，不然用户会以为所有的 model 都需要返回这些内容，有的模型在 inference 的时候可以返回更多的内容

File: docs/source/tutorials/Write_Models.md [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+
+
+## Import the model in config
+
+With ``LazyConfig System``, you can simply import the model in your config file. The following code shows how to import ``ToyModel`` in your config file:
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Write_Models.md:48]:
```suggestion
With ``LazyConfig System``, you can simply import the model in your config file. The following code shows how to use ``ToyModel`` in your config file:
```

File: docs/source/tutorials/Write_Models.md (hunk 21 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md:20]: OK, 我这边删了~

File: docs/source/tutorials/Write_Models.md (hunk 20 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md:10]: 已改~

File: docs/source/tutorials/Write_Models.md (hunk 19 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md:3]: 已改~

File: docs/source/tutorials/Write_Models.md (hunk 22 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md:42]: 已改~，我改为`__getitem__`

File: docs/source/tutorials/Write_Models.md (hunk 24 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md:48]: 已改~

File: docs/source/tutorials/Write_Models.md (hunk 23 above)
Comment (rentainhe) [reply docs/source/tutorials/Write_Models.md:43]: OK，已修改~

**Review by Ldpe2G — APPROVED** [review]
